test/Microsoft.AspNetCore.Server.Kestrel.Performance/AsciiBytesToStringBenchmark.cs

```diff
@@ -0,0 +1,492 @@
+                    break;
+            }
+
+            Verify();
+        }
+
+        [Benchmark(Baseline = true, OperationsPerInvoke = Iterations)]
+        public unsafe string EncodingAsciiGetChars()
+        {
+            for (uint i = 0; i < Iterations; i++)
+            {
+                fixed (byte* pBytes = &_asciiBytes[0])
+                fixed (char* pString = _asciiString)
+                {
+                    Encoding.ASCII.GetChars(pBytes, _asciiBytes.Length, pString, _asciiBytes.Length);
+                }
+            }
+
+            return _asciiString;
+        }
+
+        [Benchmark(OperationsPerInvoke = Iterations)]
+        public unsafe byte[] KestrelBytesToString()
+        {
+            for (uint i = 0; i < Iterations; i++)
+            {
+                fixed (byte* pBytes = &_asciiBytes[0])
+                fixed (char* pString = _asciiString)
+                {
+                    TryGetAsciiString(pBytes, pString, _asciiBytes.Length);
+                }
+            }
+
+            return _asciiBytes;
+        }
+
+        [Benchmark(OperationsPerInvoke = Iterations)]
+        public unsafe byte[] KestrelBytesToStringVectorCheck()
+        {
+            for (uint i = 0; i < Iterations; i++)
+            {
+                fixed (byte* pBytes = &_asciiBytes[0])
+                fixed (char* pString = _asciiString)
+                {
+                    TryGetAsciiStringVectorCheck(pBytes, pString, _asciiBytes.Length);
+                }
+            }
+
+            return _asciiBytes;
+        }
+
+        [Benchmark(OperationsPerInvoke = Iterations)]
+        public unsafe byte[] KestrelBytesToStringVectorized()
+        {
+            for (uint i = 0; i < Iterations; i++)
+            {
+                fixed (byte* pBytes = &_asciiBytes[0])
+                fixed (char* pString = _asciiString)
+                {
+                    TryGetAsciiStringVectorized(pBytes, pString, _asciiBytes.Length);
+                }
+            }
+
+            return _asciiBytes;
+        }
+
+        public static unsafe bool TryGetAsciiStringVectorized(byte* input, char* output, int count)
+        {
+            // Calcuate end position
+            var end = input + count;
+            // Start as valid
+            var isValid = true;
+
+            if (Vector.IsHardwareAccelerated && count >= Vector<sbyte>.Count)
+            {
+                // Jump forward to vector code to reduced cost when smaller than vector size
+                // - by default, forward jumps are predicted non-taken and near code is hotter in instruction decode
+                // When larger than Vector, vectorization will be able to make the cost back
+                goto Vectorized;
```

Comment:
Nogoto performs about same as with goto except on very short strings like KeepAlive; but still better than current, so have gone for that as the implementation